### ISSUE TITLE
moving window between monitors or resizing mixes up previous lines

### ISSUE THREAD

**Issue by thomasjachmann** [kovidgoyal/kitty#242]
Got this on a Retina MacbookPro (2016) running macOS Sierra (10.12.6).

When you move a window between different monitors (in my case a hidpi and a lowdpi monitor), the prompt line is repeated. Sometimes, the cursor is on the second but last line then and otherwise the last lines above the prompt vanish one by one.

See attached screenshots, the sequence starts on the lowdpi screen and then just moves between the two screens. The font size changes (see #77), but this also happens with the code from the dpi branch that fixes that.

![initial](https://user-images.githubusercontent.com/3219/34204161-007ed0a4-e57e-11e7-9a6b-1ac3876a7769.png)
<img width="435" alt="moved to retina" src="https://user-images.githubusercontent.com/3219/34204168-06848cb4-e57e-11e7-9fa2-892fff458d0b.png">
![moved to non-retina](https://user-images.githubusercontent.com/3219/34204171-097f6d9e-e57e-11e7-8dbf-56df55b05db1.png)
<img width="435" alt="moved to retina" src="https://user-images.githubusercontent.com/3219/34204190-163c46c4-e57e-11e7-85b0-2ef93de79a19.png">
![moved to non-retina](https://user-images.githubusercontent.com/3219/34204197-199685be-e57e-11e7-829a-a9c4fef4df04.png)
<img width="435" alt="moved to retina" src="https://user-images.githubusercontent.com/3219/34204204-1c5a8a02-e57e-11e7-8089-2795631ef264.png">

When resizing the window, similar things happen.

I'd expect the cursor line not to multiply and all previously shown line still being available in the window/scrollback. Also, the scrolling shouldn't change when moving a window to another monitor.

**Comment by kovidgoyal** [comment]
I cannot make this happen when resizing windows. I of course, dont have a multi monitor setup to test dpi changes, but the mechanism is identical in both cases. 

The reason it happens is that after a resize, your shell will clear and redraw whatever line the cursor is on. kitty uses various heuristics to try to position the cursor on the correct line after a resize, you can see various tests for it in kitty_tests/screen.py (test_resize and test_cursor_after_resize). If you can propose a test that matches what you are doing, I will look into fixing it. 

Basically, in test_cursor_resize, create a scenario that matches what you are doing. Something like:

```py
s = self.create_screen(cols=20, lines=10)  # change to match
for n in '1234567':
    s.draw(n), s.linefeed(), s.carriage_return()
s.resize(whatever)
# test to see if the cursor is on the correct line after resizing.
```

**Comment by kovidgoyal** [comment]
Oh and you can run the test by `./test.py cursor_after_resize`

**Comment by kovidgoyal** [comment]
I suppose the best way to fix this would be to get rid of the heuristics entirely and instead have the resize code track cursor position during rewrapping. I'll look into how feasible that is.

**Comment by thomasjachmann** [comment]
@kovidgoyal I never got around to create the tests, sorry. Thanks for the solution, this works perfectly now!

**Comment by bradlyatc** [comment]
so far so good here, Thanks!
